### DISCUSSION TITLE
Logic in the `consumers.py` instead of `views.py`?

### DISCUSSION THREAD

**Discussion by mikbuch** [django/channels#1801] [Q&A] [answered]
Hi, I just want to ask about the good practices about where to put the logic of my application. Normally, in Django, the logic is handled in `views.py`. In my application I have a case where my application just receives messages over a WebSocket, calculates something, and sends the messages further, again over the WebSocket. For that I don't really need to handle HTTP requests, so theoretically there is no need to put that part of the logic in `views.py`. Unless there are some reasons, arguments, or good practices, or rules that do not allow me to put my logic in the `consumers.py`. Another thing to mention is that the calculation (the logic) is triggered by an incoming message, and that it has to process each message.

Looking forward for hearing your opinions :)

Best,
Mikolaj

**Comment by carltongibson** [comment, marked as answer]
Hi @mikbuch — no harm in putting logic in `consumers.py`, especially to begin. 

Eventually you want to break out the _app-specific_ logic from the _handling requests_ logic, so you might make methods to encapsulate that, which could be called from your consumers or your views, or backend scripts even.

I don't think there's a one-sized answer for this kind of thing…

**Reply by mikbuch** [reply]
Hi @carltongibson, thank you for your answer!

My team and I have just finished configuring WebSocket connections in the `consumers.py`, and we are about to put the logic in the application. I have slowly began to realize that, in fact, there is nothing wrong in putting the logic in the `consumers.py` — your answer assured me of this notion, thanks!

After the logic will be there, I will leave a reply here (or a comment) with the summary of our experiences of calling logic methods from inside the `consumers.py`.

**Reply by carltongibson** [reply]
Cool. Have fun!

**Comment by mikbuch** [comment]
Just to leave an account here: we have implemented the logic in the `consumers.py` and everything works fine. We don't experience any performance issues on the backend side.
